tests/NHSD.BuyingCatalogue.Identity.Api.UnitTests/Services/PasswordServiceTests.cs

```diff
@@ -0,0 +1,240 @@
+                    MockUserManager.Object);
+
+                await service.SendResetEmailAsync(new ApplicationUser(), null);
+            }
+
+            Assert.ThrowsAsync<ArgumentNullException>(SendResetEmailAsync);
+        }
+
+        [TestCase("")]
+        [TestCase("\t")]
+        public void SendResetEmailAsync_EmptyOrWhiteSpaceCallback_ThrowsException(string callback)
+        {
+            async Task SendResetEmailAsync()
+            {
+                var service = new PasswordService(
+                    Mock.Of<IEmailService>(),
+                    new PasswordResetSettings(),
+                    MockUserManager.Object);
+
+                await service.SendResetEmailAsync(new ApplicationUser(), callback);
+            }
+
+            Assert.ThrowsAsync<ArgumentException>(SendResetEmailAsync);
+        }
+
+        [Test]
+        public async Task SendResetEmailAsync_SendsExpectedEmail()
+        {
+            const string emailAddress = "a@b.com";
+            const string expectedCallback = "https://identity/account/resetPassword?token=1234&emailAddress=a@b.com";
+            var expectedUser = new ApplicationUser
+            {
+                Email = emailAddress,
+                FirstName = "Eggs",
+                LastName = "Benedict",
+            };
+
+            var expectedSender = new EmailAddress("Uncle Robert", "uncle@bob.com");
+            const string expectedSubject = "Password Reset";
+
+            var configuredMessage = new EmailMessage
+            {
+                Sender = expectedSender,
+                Subject = expectedSubject,
+                HtmlBody = EmailMessage.ResetPasswordLinkPlaceholder,
+                TextBody = EmailMessage.ResetPasswordLinkPlaceholder,
+            };
+
+            var emailCount = 0;
+            EmailMessage actualEmailMessage = null;
+
+            void EmailCallback(EmailMessage message)
+            {
+                emailCount++;
+                actualEmailMessage = message;
+            }
+
+            var mockEmailService = new Mock<IEmailService>();
+            mockEmailService.Setup(e => e.SendEmailAsync(It.IsNotNull<EmailMessage>()))
+                .Callback<EmailMessage>(EmailCallback);
+
+            var settings = new PasswordResetSettings { EmailMessage = configuredMessage };
+            var service = new PasswordService(
+                mockEmailService.Object,
+                settings,
+                MockUserManager.Object);
+
+            await service.SendResetEmailAsync(expectedUser, expectedCallback);
+
+            emailCount.Should().Be(1);
+
+            actualEmailMessage.Should().NotBeNull();
+            actualEmailMessage.Sender.Should().Be(expectedSender);
+            actualEmailMessage.Recipient.Should().BeEquivalentTo(
+                new EmailAddress(expectedUser.DisplayName, expectedUser.Email));
+
+            actualEmailMessage.Subject.Should().Be(expectedSubject);
```

Comment:
Hmm.

I was looking at this, and remembering the good old days. I quite like this way of verifying calls, but there's the mockEmailService.Verify() which is supposed to be used for this purpose. I had a go at writing one though, and it looked way more ugly, so maybe this is cool?

Reply:
I agree about `Verify()` being ugly here, which is why I went for the callback-based approach.

The other option might have been to implement a custom `IEqualityComparer<T>` for tests. However, I wasn't convinced that would be worthwhile for one test.